### PR TITLE
Manage ftp.buildbot.net with Ansible

### PR DESCRIPTION
**Pull request by djmitche**
This switches from Apache to Nginx, just because we already have a handy nginx role.  Its indexes are a little less pretty,  but they'll do.

This has full support for STARTTLS in vsftpd and for https in nginx.

It also supports uploads from buildslaves.

### CODE REVIEW

File: group_vars/ftp [hunk 1]
```diff
@@ -0,0 +1,2 @@
+internal_ip: 192.168.80.243
```
Comment (sa2ajj) [inline group_vars/ftp]: We really need to go through #55.

File: jail-ftp.yml [hunk 2]
```diff
@@ -1,10 +1,30 @@
...
+  - role: ftp
+    ftp_root: "{{ ftp_root }}"
+    ftp_hostname: "{{ ftp_hostname }}"
+  # run a specialized SSH daemon to receive uploads from other internal hosts
+  - role: ssh
+    sshd_port: 2200
+    listen_address: "{{ internal_ip }}"
+    extra_config: |
+        Match User buildbot
+            ChrootDirectory /data/ftp/
```
Comment (sa2ajj) [inline jail-ftp.yml]: can we use `{{ ftp_root }}` from above here?

File: roles/ftp/meta/main.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  nginx_template: static
+  server_name: "{{ ftp_hostname }}"
+  server_root: "{{ ftp_root }}"
+  extra_config_from: "roles/ftp/files/ftp-nginx.conf"
+  ssl: True
+- role: vsftp
+  server_name: "{{ ftp_hostname }}"
+  server_root: "{{ ftp_root }}"
+  ssl: True
+
```
Comment (sa2ajj) [inline roles/ftp/meta/main.yml]: You really like trailing empty lines, do you not? :)

Comment (djmitche) [inline roles/ftp/meta/main.yml]: Hah, when I don't have pep8 to kill them, yes :)

File: roles/nginx/templates/static [hunk 4]
```diff
@@ -22,3 +22,59 @@ server {
...
+ 
+    # OCSP Stapling ---
+    # fetch OCSP records from URL in ssl_certificate and cache them
+    ssl_stapling off;
+    ssl_stapling_verify off;
+    ## verify chain of trust of OCSP response using Root CA and Intermediate certs
+    #ssl_trusted_certificate /path/to/root_CA_cert_plus_intermediates;
+    #resolver <IP DNS resolver>;
+
+    access_log {{nginx_log_dir}}/{{server_name}}/access.log;
```
Comment (sa2ajj) [inline roles/nginx/templates/static]: So we keep the logs for ssl and non-ssl access/error in the same file?

File: roles/vsftp/handlers/main.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+- name: restart vsftpd
+  service:
+    name: vsftpd
+    state: restarted
+
```
Comment (sa2ajj) [inline roles/vsftp/handlers/main.yml]: And sometimes even 2? :)

File: roles/nginx/tasks/main.yml [hunk 6]
```diff
@@ -36,6 +36,29 @@
...
+    state: directory
+  when: ssl
+
+- name: create ssl cert file
+  copy:
+    content: "{{ ssl_keys[server_name]['cert'] }}"
+    dest: "{{ nginx_cert_dir }}/{{ server_name }}.crt"
+  when: ssl
+
+- name: create ssl key file
```
Comment (sa2ajj) [inline roles/nginx/tasks/main.yml]: permissions?

File: group_vars/ftp [hunk 7]
```diff
@@ -0,0 +1,2 @@
...
+
```
Comment (sa2ajj) [inline group_vars/ftp]: empty line :)

File: roles/vsftp/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    src: vsftpd.conf.j2
+    dest: "/usr/local/etc/vsftpd.conf"
+  notify: restart vsftpd
+
+- name: start vsftpd
+  service:
+    name: vsftpd
+    enabled: true
+    state: running
+
```
Comment (sa2ajj) [inline roles/vsftp/tasks/main.yml]: empty line

**Comment by sa2ajj** [comment]
We can declare that disk space is cheap (to keep all those trailing blank lines) and merge this :)

I like it.

**Comment by djmitche** [comment]
followed-on with 6e8171efa42043c589b199ff9370cc3ae456db7e and 23b1762bdb0b693b81a610a9f69d2c9bc5b476cb

**Pull request merged**